torchaudio/datasets/tedlium.py

```diff
@@ -0,0 +1,158 @@
+        "data_path": "",
+        "subset": "train",
+    },
+    "release3": {
+        "folder_in_archive": "TEDLIUM_release-3",
+        "url": "http://www.openslr.org/resources/51/TEDLIUM_release-3.tgz",
+        "checksum": "ad1e454d14d1ad550bc2564c462d87c7a7ec83d4dc2b9210f22ab4973b9eccdb",
+        "data_path": "data/",
+        "subset": None,
+    },
+}
+
+Tedlium_item = namedtuple(
+    "Tedlium_item", ["waveform", "sample_rate", "transcript", "talk_id", "speaker_id", "identifier"]
+)
+
+
+class TEDLIUM(Dataset):
+    """
+    Create a Dataset for Tedlium. Each item is a tuple of the form:
+    [waveform, sample_rate, transcript, talk_id, speaker_id, identifier]
+    """
+
+    def __init__(
+        self, root: str, release: str = RELEASE, subset: str = None, download: bool = False, audio_ext=".sph"
+    ) -> None:
+        """Constructor for TEDLIUM dataset
+
+        Args:
+            root (str): Path containing dataset or target path where its downloaded if needed
+            release (str, optional): TEDLIUM identifier (release1,release2,release3). Defaults to RELEASE.
+            subset (str, optional): Subset of data(train,test,dev) supported for release 1,2. Defaults to Train/None.
+            download (bool, optional): Download dataset in case is not founded in root path. Defaults to False.
+            audio_ext (str, optional): Overwrite audio extension when loading items. Defaults to ".sph".
+
+        Raises:
+            RuntimeError: If release identifier does not match any supported release,
+        """
+        self._ext_audio = audio_ext
+        if release in _RELEASE_CONFIGS.keys():
+            folder_in_archive = _RELEASE_CONFIGS[release]["folder_in_archive"]
+            url = _RELEASE_CONFIGS[release]["url"]
+            subset = subset if subset else _RELEASE_CONFIGS[release]["subset"]
+        else:
+            # Raise warning
+            raise RuntimeError(
+                "The release {} does not match any of the supported tedlium releases{} ".format(
+                    release, _RELEASE_CONFIGS.keys(),
+                )
+            )
+
+        basename = os.path.basename(url)
+        archive = os.path.join(root, basename)
+
+        basename = basename.split(".")[0]
+
+        self._path = os.path.join(root, folder_in_archive, _RELEASE_CONFIGS[release]["data_path"])
+        if subset in ["train", "dev", "test"]:
+            self._path = os.path.join(self._path, subset)
+        if download:
+            if not os.path.isdir(self._path):
+                if not os.path.isfile(archive):
+                    checksum = _RELEASE_CONFIGS[release]["checksum"]
+                    download_url(url, root, hash_value=checksum)
+                extract_archive(archive)
+
+        walker = walk_files(self._path, suffix=".stm", prefix=False, remove_suffix=True)
+        self._walker = list(walker)
+        self._extended_walker = []
```

Comment:
I'm not sure I understand why we need an extended walker in place of updating walker? It seems like this simply extracts information from a file. Could this be done in load item function?

Reply:
It does simple extract information for the files contained in the original walker. I create it because we need this information before calling the load item function so the dataset contains samples equals to total lines in all files (1 line is one sentence) instead of containing samples equals to files.

This dataset is a bit tricky because each file contains a full ted talk and no one in speech recognition (where this dataset is usually used) treats a full talk as a training sample. We split the file and audio file into multiple samples (sentences).

This means that once we instantiate the dataset we need to know how many samples we have(lines in all files) so the user can index any sample N at any time or shuffle it. If we were to implement this on the load function then this functionality will be way harder. See the original post where I explain with detail how the files are iterated.

I'm happy to discuss which method is the best method to achieve this functionality but I'm checking first if you understand why this extender walker is here first, maybe I didn't explain it correctly on the original post

Reply:
Thanks for adding details :) What I meant is more about the particular use of `_walker` and `_extended_walker`. I would have suggested something like:
```python
        self._walker = []
        for file in walk_files(self._path, suffix=".stm", prefix=False, remove_suffix=True):
            stm_path = os.path.join(self._path, "stm", file + ".stm")
            with open(stm_path) as f:
                l = len(f.readlines())
                self._walker.extend((file, line) for line in range(l))
```
or, to stick with generators,
```python
    def _generate_walker(...):
        for file in walk_files(self._path, suffix=".stm", prefix=False, remove_suffix=True):
            stm_path = os.path.join(self._path, "stm", file + ".stm")
            with open(stm_path) as f:
                for line in range(len(f.readlines())):
                    yield (file, line)

    self._walker = list(_generate_walker(...))
```

Also, why just indexing and not storing the line directly since the file is already being read? How big is the total text?
```python
    def _generate_walker(...):
        for file in walk_files(self._path, suffix=".stm", prefix=False, remove_suffix=True):
            stm_path = os.path.join(self._path, "stm", file + ".stm")
            with open(stm_path) as f:
                for line in f.readlines():
                    yield (file, line)

    self._walker = list(_generate_walker(...))
```

Reply:
Oh I see what you mean now, yes it will be cleaner to stick to the _walker name for the resultant walker.

I went ahead and checked the size of the walker list for release3 (the biggest)
```python
import numpy as np
np.array(test_dataset._extended_walker).nbytes
```
At its `316733528` byte,~316 megabytes` storing the full line and `20449080 ~20 megabytes`. I would say is a considerable size but it does remove a I/O operation each time we load the item. Does any other dataset take this much space in memory?